Added monitoring of disk space via monit

phdoerfler opened the pull request:
This fixes https://github.com/r-raymond/nixos-mailserver/issues/55.

r-raymond commented on mail-server/monit.nix:
```diff
@@ -0,0 +1,38 @@
+      enable = true;
+      config = ''
+        set daemon 120
+        set alert root@localhost
```
this should be configurable. Most people running a webserver will not bother checking the root account.

phdoerfler replied:
Yes it should and I believe I suggested exactly that somewhere but you're right, I might as well implement it in this PR.
This being said I believe you can redirect mails to root to some other account in the postfix config. Doing that would allow all the other things in the system that hardcoded send stuff to root to work, too.

r-raymond commented on mail-server/monit.nix:
```diff
@@ -0,0 +1,38 @@
+        set alert root@localhost
+        set httpd port 2812 and use address localhost
+            allow localhost
+            allow admin:obwjoawijerfoijsiwfj29jf2f2jd
```
what is this?

phdoerfler replied:
this is what I believe to be the minimal config monit requires to work.
The funny looking line at the end is the password for the user "admin" using HTTP Basic Auth.
In the end I doubt it matters much what is written there since the server is not reachable from outside of localhost. But it doesn't hurt to have some strong password there because the monit tools themselves don't require you to input the password (e.g. `monit report`).

r-raymond commented on mail-server/monit.nix:
```diff
@@ -0,0 +1,38 @@
+            allow localhost
+            allow admin:obwjoawijerfoijsiwfj29jf2f2jd
+        
+        check device root with path /
```
We should check the path of vmailDirectory here. I e.g. like to put my mail directory on a different partition.

phdoerfler replied:
Good idea! I'll add that.

phdoerfler commented:
I haven't lost track of this but can't work on it because my nixos server believes it needs to compile rustc and is failing horribly. This is apparently a known problem and being worked on. But until it is solved I can't run `nixos-rebuild switch` so working on pull requests is a bit adventurous ;)

phdoerfler commented:
I decided to just hand the entire config to the user, albeit with a already very useful example config.
The reason for this is that what people want to be warned about depends highly on the people and on the infrastructure the mail server is running on.
I started looking into adding own little options for things like "special case: vmail is on a different file system" but in the end this would complicate things more than necessary. For instance: You would first have to identify what device or filesystem is holding the vmail directory and then check that particular device / filesystem. You can't just tell Monit to check if the FS holding the vmail directory has free space left or not. And so on. You get the idea.

My only regret (and that's why I decided to not enable it by default) is that you still have to pick a mail address to send all these alerts to. If we had an extra option for `importantMailGoesHere = "root@localhost"` or something then I'd add that and enable it by default.

phdoerfler commented:
@r-raymond Something's afoot with the travis CI build and I doubt it being caused by this PR or the others. Can you look into it?
Curious candidates are:

- https://travis-ci.org/r-raymond/nixos-mailserver/builds/336791302#L5215
- https://travis-ci.org/r-raymond/nixos-mailserver/builds/336791302#L5348
- https://travis-ci.org/r-raymond/nixos-mailserver/builds/336791302#L5354 (and following errors)

r-raymond commented:
It seems to be related to an SSL bug in dovecot:

```
server# [  233.257819] dovecot[1130]: imap-login: Error: Failed to initialize SSL server context: Couldn't parse DH parameters: error:0906D06C:PEM routines:PEM_read_bio:no start line: Expecting: DH PARAMETERS: user=<>, rip=192.168.1.1, lip=192.168.1.2, session=<Z2g4KkRkbsjAqAEB>
server# [  233.299718] dovecot[1130]: imap-login: Disconnected: TLS initialization failed. (no auth attempts in 0 secs): user=<>, rip=192.168.1.1, lip=192.168.1.2, session=<Z2g4KkRkbsjAqAEB>
```

compare to [archlinux](https://bbs.archlinux.org/viewtopic.php?id=233452). I'll try to find a  solution

phdoerfler commented:
I am a bit stumped as to why Github fails to merge default.nix. It merges just fine on my local clone.

phdoerfler commented:
It would seem rspamd is irritated by the monitoring rules I specified:

```
$ journalctl -u rspamd -xe

Feb 14 14:50:03 turtle rspamd[1120]: <31f79f>; task; rspamd_worker_error_handler: abnormally closing connection from: /run/rspamd/rspamd.sock, error: IO read error: unexpected EOF
Feb 14 14:50:03 turtle rspamd[1120]: <31f79f>; task; rspamd_worker_error_handler: abnormally closing connection from: /run/rspamd/rspamd.sock, error: IO write error: Broken pipe
Feb 14 14:52:08 turtle rspamd[1120]: <daf07b>; task; accept_socket: accepted connection from /run/rspamd/rspamd.sock port 0, task ptr: 0000000001B6E5A0
Feb 14 14:52:08 turtle rspamd[1120]: <daf07b>; task; rspamd_worker_error_handler: abnormally closing connection from: /run/rspamd/rspamd.sock, error: IO read error: unexpected EOF
Feb 14 14:52:08 turtle rspamd[1120]: <daf07b>; task; rspamd_worker_error_handler: abnormally closing connection from: /run/rspamd/rspamd.sock, error: IO write error: Broken pipe
Feb 14 14:54:09 turtle rspamd[1120]: <2b2242>; task; accept_socket: accepted connection from /run/rspamd/rspamd.sock port 0, task ptr: 0000000001B6E800
Feb 14 14:54:09 turtle rspamd[1120]: <2b2242>; task; rspamd_worker_error_handler: abnormally closing connection from: /run/rspamd/rspamd.sock, error: IO read error: unexpected EOF
Feb 14 14:54:09 turtle rspamd[1120]: <2b2242>; task; rspamd_worker_error_handler: abnormally closing connection from: /run/rspamd/rspamd.sock, error: IO write error: Broken pipe
```

You can clearly see the 2 minute interval I specified in the default configuration.
The following lines from my monit default config use the rspamd unix socket to check if it's still alive. Apparently rspamd is irritated by that, hence the logs:

```
check process rspamd with pidfile /var/run/rspamd.pid
    start program = "${pkgs.systemd}/bin/systemctl start rspamd"
    stop program = "${pkgs.systemd}/bin/systemctl stop rspamd"
    if failed unixsocket /var/run/rspamd/rspamd.sock for 5 cycles then restart
```

I am going to look into this before this PR gets merged.

phdoerfler commented:
I have removed the check for the unix socket.
monit is apparently capable of sending a custom message through that unix socket that would satisfy rspamd, however I don't feel like conjuring up something like that. I don't think it's worth it, hence I removed the check alltogether.

jbboehr commented:
Would it make sense to make `alertAddress` an array or recommend that it be set to an external email address? Does `set mailserver
              localhost` cause it to be sent through localhost? I assume that's undesirable if the local server is down.

r-raymond commented:
@jbboehr do I understand your question correctly that you'd like to have multiple alert addresses? I think this is more of a "you are running low on space on your harddrive" type of monitoring, not a "your server went down do something about it" type of monitoring. Not saying that the latter is not something we could think about, but the former is certainly the more common case.

jbboehr commented:
@r-raymond Ah I see. I saw `check process postfix` and `check process dovecot` and assumed it was for monitoring the availability of the email server itself. Presumably the alert message will not be deliverable if either of these checks fail (or maybe just the dovecot one)?
